src/QsCompiler/Transformations/CallGraph/CallGraphBase.cs

```diff
@@ -0,0 +1,255 @@
+            }
+
+            this.FromCallableName = fromCallableName;
+            this.ToCallableName = toCallableName;
+            this.ReferenceRange = referenceRange;
+        }
+
+        /// <summary>
+        /// Determines if the object is the same as the given edge, ignoring the
+        /// ordering of key-value pairs in the type parameter dictionaries.
+        /// </summary>
+        public bool Equals(CallGraphEdgeBase edge) =>
+            this.FromCallableName.Equals(edge.FromCallableName)
+            && this.ToCallableName.Equals(edge.ToCallableName)
+            && this.ReferenceRange.Equals(edge.ReferenceRange);
+    }
+
+    /// <summary>
+    /// Base class for call graph node types.
+    /// </summary>
+    public abstract class CallGraphNodeBase : IEquatable<CallGraphNodeBase>
+    {
+        /// <summary>
+        /// The name of the represented callable.
+        /// </summary>
+        public QsQualifiedName CallableName { get; }
+
+        /// <summary>
+        /// Base constructor for call graph nodes. Initializes CallableName.
+        /// Throws an ArgumentNullException if argument is null.
+        /// </summary>
+        protected CallGraphNodeBase(QsQualifiedName callableName)
+        {
+            if (callableName is null)
+            {
+                throw new ArgumentException(nameof(callableName));
+            }
+
+            this.CallableName = callableName;
+        }
+
+        /// <inheritdoc/>
+        public override bool Equals(object obj)
+        {
+            return obj is CallGraphNodeBase && this.Equals((CallGraphNodeBase)obj);
+        }
+
+        /// <inheritdoc/>
+        public bool Equals(CallGraphNodeBase other) =>
+            this.CallableName.Equals(other.CallableName);
+
+        /// <inheritdoc/>
+        public override int GetHashCode()
+        {
+            HashCode hash = default;
+            hash.Add(this.CallableName);
+            return hash.ToHashCode();
+        }
+    }
+
+    /// <summary>
+    /// Base class for call graph types.
+    /// </summary>
+    public abstract class CallGraphBase<TNode, TEdge>
+        where TNode : CallGraphNodeBase
+        where TEdge : CallGraphEdgeBase
+    {
+        // Static Elements
+
+        /// <summary>
+        /// Returns an empty dependency for a node.
+        /// </summary>
+        protected static ILookup<TNode, TEdge> EmptyDependency() =>
+            ImmutableArray<KeyValuePair<TNode, TEdge>>.Empty
+            .ToLookup(kvp => kvp.Key, kvp => kvp.Value);
+
+        // Member Fields
+
+        /// <summary>
+        /// This is a dictionary mapping source nodes to information about target nodes. This information is represented
+        /// by a dictionary mapping target node to the edges pointing from the source node to the target node.
+        /// </summary>
+        protected readonly Dictionary<TNode, Dictionary<TNode, ImmutableArray<TEdge>>> dependencies =
```

Comment:
Seems like it might be cleaner to make this private and only expose it via `GetDirectDependencies`. Otherwise, this field is effectively public, which exposes an implementation detail about the call graph that I'm not sure was intended.

Reply:
I especially like this idea! 😃

Reply:
done